core-java/src/main/java/org/spine3/server/procman/ProcessManager.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright 2015, TeamDev Ltd. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.server.procman;
+
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.protobuf.Message;
+import com.google.protobuf.Timestamp;
+import org.spine3.base.*;
+import org.spine3.internal.EventHandlerMethod;
+import org.spine3.server.Entity;
+import org.spine3.server.internal.CommandHandlerMethod;
+import org.spine3.util.Classes;
+import org.spine3.util.Events;
+import org.spine3.util.MethodMap;
+
+import javax.annotation.CheckReturnValue;
+import javax.annotation.Nullable;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A central processing unit used to maintain the state of the sequence and determine the next processing step
+ * based on intermediate results.
+ * <p>
+ * An independent component that reacts to domain events in a cross-aggregate, eventually consistent manner.
+ * Used for externalizing the decisions on the logic flow from the business logic.
```

Comment:
OK, I think it looks like the business logic, too. But I've read from different sources that PMs don't perform any business logic. E.g., from [CQRS Journey Guide](https://msdn.microsoft.com/en-us/library/jj591569.aspx):
```
It's important to note that the process manager does not perform any business logic.  
It only routes messages, and in some cases translates between message types.  
You should not use a process manager to implement any business logic in your domain.  
Business logic belongs in the aggregate types.
```
But they also state that PMs are used for business processes. 
So, as I understand, they call messages routing "business processes" (though it is related to the domain and I think it can be called "business logic"), maybe, to separate PMs from ARs etc. So that nobody will perform in PMs any logic except messages routing.